torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -156,7 +156,8 @@ void RpcAgent::rpcRetryCallback(
     // just the retry count.
     LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_ + 1)
               << " failed";
-    if (earliestRpc->retryCount_ < earliestRpc->options_.maxRetries) {
+    if (earliestRpc->retryCount_ < earliestRpc->options_.maxRetries &&
+        rpcAgentRunning_.load()) {
```

Comment:
as we don't hold lock here, this is best_effort check? Could it happen that `rpcAgentRunning_` toggle to false immediately after this check? Will that cause any issues?

Reply:
Yes, this is a best-effort check, basically as a guard against scheduling retries if the agent has been shutdown. If `rpcAgentRunning_` is set to False immediately after this:
1. If the retry thread is processing retries, #35263 will ensure that performing sends when the agent is shutdown is handled gracefully.
1. If the retry thread is not processing retries (sleeping), then the retry function will simply return upon waking up and not process any further retries (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rpc_agent.cpp#L85).